src/schema/schema_create.c

```diff
@@ -322,6 +322,60 @@ __wt_schema_index_source(WT_SESSION_IMPL *session,
 }
 
 /*
+ * __fill_index --
+ *	Fill the index from the current contents of the table.
+ */
+static int
+__fill_index(WT_SESSION_IMPL *session, WT_TABLE *table, const char *name)
+{
+	WT_DECL_RET;
+	WT_CURSOR *tcur, *icur, *child;
+	WT_CURSOR_INDEX *cindex;
+	WT_CURSOR_TABLE *ctable;
+	WT_INDEX *idx;
+	WT_SESSION *sess;
+
+	sess = (WT_SESSION *)session;
+	tcur = NULL;
+	icur = NULL;
+	WT_RET(__wt_schema_open_colgroups(session, table));
+
+	/*
+	 * If the column groups have not been completely created,
+	 * there cannot be data inserted yet, and we're done.
+	 */
+	if (!table->cg_complete)
+		return (0);
+
+	WT_ERR(sess->open_cursor(sess, table->name, NULL, "readonly", &tcur));
+	WT_ERR(sess->open_cursor(sess, name, NULL, NULL, &icur));
+
+	ctable = (WT_CURSOR_TABLE *)tcur;
+	cindex = (WT_CURSOR_INDEX *)icur;
+	child = cindex->child;
+	idx = cindex->index;
+
+	while ((ret = tcur->next(tcur)) == 0) {
+		/*
+		 * For LSM indices, we must drop locks so that the
+		 * LSM worker thread can grab the schema lock during
+		 * a switch.
+		 */
+		WT_WITHOUT_LOCKS(session,
+		    ret = __wt_apply_single_idx(session, idx,
+		    child, ctable, child->insert));
```

Comment:
@ddanderson, I understand the LSM issues, but can you talk me through what makes this safe?

Can checkpoint see a partially-completed index (and if so, are there recovery implications)?  Is there any chance of deadlock when reacquiring the locks?